Automattic/jetpack
# CRM: 3417 - Remove invalid statuses from listview filters

Resolves Automattic/zero-bs-crm#3417 - Remove invalid statuses from listview filters

## Proposed changes:

The issue at hand: no matter what statuses were configured in settings, the default object statuses from CRM init were always available to select. For contacts, that meant one couldn't get rid of "Lead" or "Customer", particularly annoying if using a non-English system.

**For the reviewer (thanks and apologies in advance):**
* I chopped some unrelated pre-DAL3 code in 7731858e5ebc9ad9fda2b66d9db3779821dd4c4c. The bulk of the changes are in `includes/ZeroBSCRM.DAL.Fields.php`, a file we never include in our code (it's there for legacy reasons to collect some strings for translation). I recommend reviewing that commit alone first (or altogether skipping it).
* The other massive chop is in 5bf3625b4a0715a9efb7435286b5270698efaecd, removing the legacy listview filter code. I'd recommend reviewing this last (or skipping it if you want).
* With the above in mind, I've organized the commits in logical order, so hopefully that'll make the review easier.

The listview filter system was set up with lots of globals and a fair amount of repetitive code, and had "defaults" vs. "all" settings due to the pre-Emerald system of showing/hiding available filters.

Given we no longer use this UI, and the general bad state of the code, I've reimplemented listview filters. The general summary:

* All listview filters are registered to `$zbs->listview_filters`.
* The `jpcrm_listview_filters` filter is available. It's applied during `$zbs->postInit()`, just after the `jpcrm_post_init` action is called. This should be sufficiently late enough to register any filters needed.

The array is structured like this:

```
[
    ZBS_TYPE_CONTACT: [
        cat1: [
            slug1: translated_label1,
            slug2: translated_label2,
        ],
        cat2: [
            slug1: translated_label1,
            slug2: translated_label2,
        ],
    ],
    ZBS_TYPE_INVOICE: [
        cat1: [
            slug1: translated_label1,
            slug2: translated_label2,
        ],
        cat2: [
            slug1: translated_label1,
            slug2: translated_label2,
        ],
    ],
]
```

Here's an example:
```
$listview_filters[ ZBS_TYPE_CONTACT ]['general']['assigned_to_me'] = __( 'Assigned to me', 'zero-bs-crm' );
```

The category slugs are translated when rendering the listview filters, using a lookup array stored as `$quickfilter_categories`. These are the currently-defined categories:

```
$quickfilter_categories = array(
    'general' => __( 'General', 'zero-bs-crm' ),
    'status'  => __( 'Status', 'zero-bs-crm' ),
    'segment' => __( 'Segment', 'zero-bs-crm' ),
);
```

A few notes:
* Category order is currently "first registered, first displayed".
* If a provided category isn't found, it falls back to using the slug.
* If there is only one category registered, it will not display the category name. If there are more than one registered, it will show each category name.
* This change affects Mail Campaigns; no error will be generated, but its one quick filter will not show until updated.
* I noticed that there was no listview filter for quote drafts, so I added one.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






1. Install WooCommerce and MailPoet, and enable those modules.
2. Change your default contact statuses to `a,b,c,d`: `/wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=fieldoptions`
3. Change your default company statuses to `e,f,g,h`: `/wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=companies`
4. Change your default transaction statuses to `i,j,k,l`: `/wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=transactions`
5. Add a few segments.
6. Verify the listview filters work as expected.

Woo adds a filter to contacts, transactions, and invoices. MailPoet adds one to contacts. Statuses and Segments are auto-added if enabled in settings: `/wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=fieldoptions`

In `trunk`, the listview filters will show the default statuses that you've removed as well as the new ones:

![image](https://github.com/Automattic/jetpack/assets/32492176/a3b02410-1075-4f5b-8cd4-1fa1abf433ac)


In the `fix/rm/3417-remove_invalid_statuses_from_listview_filters` branch (yes, there was a typo in the branch name), you'll see the expected filters nicely categorized:

![image](https://github.com/Automattic/jetpack/assets/32492176/8378cd2e-1494-4e77-a7a7-d620931ba0eb)
